lib/sdk/remote/parent.js

```diff
@@ -0,0 +1,313 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const { Cu, Ci, Cc } = require('chrome');
+const runtime = require('../system/runtime');
+
+const MAIN_PROCESS = Ci.nsIXULRuntime.PROCESS_TYPE_DEFAULT;
+
+if (runtime.processType != MAIN_PROCESS) {
+  throw new Error('Cannot use sdk/remote/parent in a child process.');
+}
+
+const { Class } = require('../core/heritage');
+const { Namespace } = require('../core/namespace');
+const { Disposable } = require('../core/disposable');
+const { omit } = require('../util/object');
+const { when } = require('../system/unload');
+const { EventTarget } = require('../event/target');
+const { emit } = require('../event/core');
+const system = require('../system/events');
+const { AttachPoint } = require('./utils');
+const options = require('@loader/options');
+const loader = require('toolkit/loader');
+
+// Chose the right function for resolving relative a module id
+let moduleResolve;
+if (options.isNative) {
+  moduleResolve = (id, requirer) => loader.nodeResolve(id, requirer, { rootURI: options.rootURI });
+}
+else {
+  moduleResolve = loader.resolve;
+}
+// Build the sorted path mapping structure that resolveURI requires
+let pathMapping = Object.keys(options.paths)
+                        .sort((a, b) => b.length - a.length)
+                        .map(p => [p, options.paths[p]]);
+
+// Load the scripts in the child processes
+let { getNewLoaderID } = require('../../framescript/FrameScriptManager.jsm');
+let PATH = options.paths[''];
+
+const childOptions = omit(options, ['modules', 'globals']);
+const loaderID = getNewLoaderID();
+childOptions.loaderID = loaderID;
+
+const ppmm = Cc['@mozilla.org/parentprocessmessagemanager;1'].
+             getService(Ci.nsIMessageBroadcaster);
+const gmm = Cc['@mozilla.org/globalmessagemanager;1'].
+            getService(Ci.nsIMessageBroadcaster);
+
+const ns = Namespace();
+
+let processMap = new Map();
+
+function processMessageReceived({ target, data }) {
+  if (data.loaderID != loaderID)
+    return;
+  emit(this.port, ...data.args);
+}
+
```

Comment:
nit can we add some description about what a `Process` is here please, something like `// Process is a child process which contains one to many frames/tabs`.